### PR TITLE
able to change attackTick

### PR DESCRIPTION
**Pull request by djdisodo**
## Introduction
i'm making a gun plugin.
attack() doesn't worked on some guns like shoutgun(it spawns many projectiles)

## Tests
tested

### CODE REVIEW

**Comment by Thunder33345** [comment]
what about my plugin who use `CAUSE_CUSTOM` and i want attack ticks
you would break my plugin!!!!!!!

**Comment by djdisodo** [comment]
@Thunder33345 
https://github.com/pmmp/PocketMine-MP/compare/master...djdisodo:patch-2?expand=1
how about this

**Comment by Thunder33345** [comment]
looks better
the default should be 10 tho?

**Comment by Muqsit** [comment]
Nice, but `attackDelay`, `getAttackDelay` and `setAttackDelay` sound more meaningful than `attackTick`.

**Comment by Thunder33345** [comment]
more like [quazi]immune tick
blame mojang naming stuff

**Comment by dktapps** [comment]
https://xkcd.com/1172/

File: src/pocketmine/entity/Living.php [hunk 1]
```diff
@@ -560,7 +560,7 @@ public function attack(EntityDamageEvent $source) : void{
 			return;
 		}
 
-		$this->attackTime = 10; //0.5 seconds cooldown
+		if($source->getCause() !== EntityDamageEvent::CAUSE_CUSTOM) $this->attackTime = 10; //0.5 seconds cooldown
```
Comment (SOF3) [inline src/pocketmine/entity/Living.php]: OCD: Use curly braces.

**Review by SOF3 — DISMISSED** [review]

File: src/pocketmine/event/entity/EntityDamageEvent.php [hunk 2]
```diff
@@ -197,4 +200,18 @@ public function canBeReducedByArmor() : bool{
...
+	 * @return int
+	 */
+	public function getAttackTick() : int{
+		return $this->attackTick;
+	}
+	
+	/**
+	 * @param int $attackTick
+	 */
+	public function setAttackTick(int $attackTick) {
```
Comment (dktapps) [inline src/pocketmine/event/entity/EntityDamageEvent.php]: this should have a `void` typehint.

**Review by dktapps — COMMENTED** [review]
To me it seems more clear to refer to it as an attack cooldown than `attackTick` - the use is more clear from the name.

File: src/pocketmine/event/entity/EntityDamageEvent.php [hunk 3]
```diff
@@ -197,4 +200,18 @@ public function canBeReducedByArmor() : bool{
...
+	 */
+	public function getAttackCooldown() : int{
+		return $this->attackCooldown;
+	}
+	
+	/**
+	 * @param int $attackTick
+	 */
+	public function setAttackCooldown(int $attackTick) : void{
+		$this->attackTick = $attackCooldown;
```
Comment (dktapps) [inline src/pocketmine/event/entity/EntityDamageEvent.php]: looks like you missed a couple of things here

**Comment by dktapps** [comment]
The only nitpick I can make here now is the fact that this will not work for non-living entities. I'm not really sure how that can be dealt with.

**Comment by djdisodo** [comment]
@SOF3 please del change request....

**Review by dktapps — APPROVED** [review]
LGTM, will look at integrating this into 3.2.

**Comment by dktapps** [comment]
Squash-merged in the referenced commit. Thanks.

**Pull request closed without merging**